walk-in-interview/src/main/java/com/google/job/data/JobsDatabase.java

```diff
@@ -44,19 +49,64 @@
      *
      * @param jobId Id for the target job post in the database.
      * @param updatedJob Updated job post.
-     * @return A future of the detailed information of the update.
+     * @return A future of document reference for the updated job post.
      * @throws IllegalArgumentException If the job id is invalid.
      */
-    public Future<WriteResult> setJob(String jobId, Job updatedJob) throws IllegalArgumentException, IOException {
+    public Future<DocumentReference> setJob(String jobId, Job updatedJob) throws IllegalArgumentException, IOException {
+        if (jobId.isEmpty()) {
+            throw new IllegalArgumentException("Job Id should be an non-empty string");
+        }
+        
         // Sets the Job with cloud firestore id and ACTIVE status
         Job job = updatedJob.toBuilder()
                 .setJobId(jobId)
                 .build();
 
-        // Overwrites the whole job post
-        return FireStoreUtils.getFireStore()
-                .collection(JOB_COLLECTION).document(jobId)
-                .set(job);
+        // Runs an asynchronous transaction
+        ApiFuture<DocumentReference> futureTransaction = FireStoreUtils.getFireStore().runTransaction(transaction -> {
+            final DocumentReference documentReference = FireStoreUtils.getFireStore()
+                    .collection(JOB_COLLECTION).document(jobId);
+
+            // Verifies if the current user can update the job post with this job id
+            // TODO(issue/25): incorporate the account stuff into job post.
+            verifyUserCanUpdateJob(jobId);
```

Comment:
this doesn't really work. You should access your db within the transaction at this time, see the link i gave you:
https://firebase.google.com/docs/firestore/manage-data/transactions
```
 DocumentSnapshot snapshot = transaction.get(docReference);
 if( snapshot == null ){ [...] // job doesn't exist }
 [...]
```
Same for the other one.